Guessing wrong content type and applying filters on binary file shouldn't prevent building of RPMs

When I commit the attached binary file, the RPM maker treats this file as plain text, tries to filter it and refuses to build a RPM with the following error message:

> [ERROR] Cannot filter file binary.png.
> [ERROR] unknown encoding: unknown-8bit

I suppose that some random bytes in my file make the rpm maker guess that it's plain text instead of a binary file (all other binary files I commit, are treaded correctly).

A solution would be to just log a warning and ignore that the filters cannot be applied, instead of just refusing to build the RPM.

Please note, the attached file is not really a PNG, but github refuses to upload arbitrary binary files.

![binary](https://f.cloud.github.com/assets/6065693/1644262/7b00a7e2-58df-11e3-88da-ad27acc9a716.png)

--- a/src/config_rpm_maker/token/tokenreplacer.py
+++ b/src/config_rpm_maker/token/tokenreplacer.py
@@ -19,6 +19,7 @@ import re
 import os
 
 from logging import getLogger
+from os.path import getsize
 
 import config_rpm_maker.magic
 
@@ -166,26 +167,44 @@ class TokenReplacer(object):
             content = content.replace("@@@%s@@@" % token_name, replacement)
             self.token_used.add(token_name)
 
+    def _read_content_from_file(self, filename):
+
+        with open(filename, "r") as input_file:
+            file_content = input_file.read()
+
+        return file_content
+
+    def _perform_filtering_on_file(self, filename, file_content, file_encoding, html_escape):
+        file_content = file_content.decode(file_encoding)
+
+        if html_escape:
+            file_content = self.html_escape_function(os.path.basename(filename), file_content)
+
+        file_content_filtered = self.filter(file_content)
+
+        with open(filename, "w") as output_file:
+            output_file.write(file_content_filtered.encode(file_encoding))
+
     def filter_file(self, filename, html_escape=False):
         try:
             self.file_size_limit = config.get('max_file_size', DEFAULT_FILE_SIZE_MAXIMUM)
-            if os.path.getsize(filename) > self.file_size_limit:
-                raise Exception("FileTooFatException : %s\n\t(size is %s bytes, limit is %s bytes)" % (os.path.basename(filename), os.path.getsize(filename), self.file_size_limit))
 
-            with open(filename, "r") as input_file:
-                file_content = input_file.read()
+            file_size = getsize(filename)
+            if file_size > self.file_size_limit:
+                raise Exception("FileTooFatException : %s\n\t(size is %s bytes, limit is %s bytes)" % (os.path.basename(filename), file_size, self.file_size_limit))
+
+            file_content = self._read_content_from_file(filename)
 
             file_encoding = self._get_file_encoding(file_content)
-            if file_encoding and file_encoding != 'binary':
-                file_content = file_content.decode(file_encoding)
-                if html_escape:
-                    file_content = self.html_escape_function(os.path.basename(filename), file_content)
+            if file_encoding:
+                if file_encoding != 'binary' and file_encoding != 'unknown-8bit':
+                    self._perform_filtering_on_file(filename, file_content, file_encoding, html_escape)
+                else:
+                    LOGGER.warn('Not filtering file "%s" since it has encoding "%s".', filename, file_encoding)
 
-                file_content_filtered = self.filter(file_content)
-                with open(filename, "w") as output_file:
-                    output_file.write(file_content_filtered.encode(file_encoding))
         except MissingTokenException as exception:
             raise MissingTokenException(exception.token, filename)
+
         except Exception as e:
             raise CannotFilterFileException('Cannot filter file %s.\n%s' % (os.path.basename(filename), str(e)))
 

--- a/test/unittests/token/tokenreplacer_test.py
+++ b/test/unittests/token/tokenreplacer_test.py
@@ -18,6 +18,7 @@
 
 import unittest
 
+from mock import Mock, patch
 
 from config_rpm_maker.token.cycle import ContainsCyclesException
 from config_rpm_maker.token.tokenreplacer import MissingTokenException, TokenReplacer
@@ -68,3 +69,18 @@ class TokenReplacerTest(unittest.TestCase):
     def test_should_determine_token_recursion(self):
         self.assertRaises(ContainsCyclesException, TokenReplacer, {"FOO": "@@@BAR@@@", "BAR": "@@@FOO@@@"})
         self.assertRaises(ContainsCyclesException, TokenReplacer, {"FOO": "@@@BAR@@@", "BAR": "@@@BLO@@@", "BLO": "@@@FOO@@@"})
+
+    @patch('config_rpm_maker.token.tokenreplacer.config')
+    @patch('config_rpm_maker.token.tokenreplacer.getsize')
+    def test_should_not_filter_file_with_encoding_unknown_8bit(self, mock_get_size, mock_config):
+
+        mock_get_size.return_value = 10
+        mock_config.get.return_value = 20
+
+        mock_token_replacer = Mock(TokenReplacer)
+        mock_token_replacer._read_content_from_file.return_value = 'fake binary file content'
+        mock_token_replacer._get_file_encoding.return_value = 'unknown-8bit'
+
+        TokenReplacer.filter_file(mock_token_replacer, "binary.file")
+
+        self.assertEqual(0, mock_token_replacer._perform_filtering_on_file.call_count)

